File: source/ts/src/core/nakshatra.ts
import { NAKSHATRA_SPAN, NAKSHATRA_PADA_SPAN, nakshatraOf } from '../utils/constants';
import type { NakshatraInfo } from '../types/elements';

export function computeNakshatraFromLongitude(
  siderealMoon: number,
  name: string,
): NakshatraInfo {
  const index = nakshatraOf(siderealMoon);
  const degreesInNakshatra = Math.max(siderealMoon - index * NAKSHATRA_SPAN, 0);
  const pada = Math.min(Math.floor(degreesInNakshatra / NAKSHATRA_PADA_SPAN) + 1, 4);
  const completionPercentage = (degreesInNakshatra / NAKSHATRA_SPAN) * 100;
  return {
    index,
    name,
    pada,
    degreesInNakshatra: Math.round(degreesInNakshatra * 10000) / 10000,
    completionPercentage: Math.round(completionPercentage * 100) / 100,
    endTime: null,
  };
}

export function getNakshatraIndexAtTime(
  date: Date,
  getCachedMoon: (d: Date) => number,
): number {
  return nakshatraOf(getCachedMoon(date));
}

File: source/go/internal/core/nakshatra.go
package core

import (
	"math"

	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/types"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/utils"
)

func ComputeNakshatraFromLongitude(siderealMoon float64, name string) types.NakshatraInfo {
	index := utils.NakshatraOf(siderealMoon)
	degreesInNakshatra := math.Max(siderealMoon-float64(float64(index)*utils.NakshatraSpan), 0)
	pada := int(math.Floor(degreesInNakshatra/utils.NakshatraPadaSpan)) + 1
	if pada > 4 {
		pada = 4
	}
	completionPercentage := (degreesInNakshatra / utils.NakshatraSpan) * 100
	return types.NakshatraInfo{
		Index:                index,
		Name:                 name,
		Pada:                 pada,
		DegreesInNakshatra:   jsnum.Round(degreesInNakshatra*10000) / 10000,
		CompletionPercentage: jsnum.Round(completionPercentage*100) / 100,
		EndTime:              nil,
	}
}

func GetNakshatraIndexAtTime(ms int64, getCachedMoon LongitudeAt) int {
	return utils.NakshatraOf(getCachedMoon(ms))
}
